src/PPU.cpp

```diff
@@ -347,48 +353,147 @@ void PPU::updateObjects() {
+
+	// Early out if current ly doesn't intersect with the window
+	int y = ly - wy;
+	if (y < 0) {
+		*startingIndex = 160;
+		return;
+	}
+
+	// Find y of the tile to work with
+	uint16_t tileY = y / 8;
+	uint16_t tileYPos = tileY * 32;
+
+	// Starting index indicates where the first position of the window is in the scanline (-7 to 159)
+	*startingIndex = wx - 7;
+
+	// Construct scanline, max 21 tiles need to be loaded
+	for (int i = 0; i < 21; i++) {
+		// Find x of the tile to work with
+		uint8_t x = i * 8;
+		uint16_t tileX = i;
+
+		// Find tile to load
+		uint16_t tileIndex = tileYPos + tileX;
+
+		uint8_t index = mmu->directRead(windowStart + tileIndex);
+		uint16_t start = (index > 127) ? secondHalfStart : firstHalfStart;
+		index &= 0x7F;
+
+		// Find the y of the correct line within the tile
+		uint8_t lineY = y - (tileY * 8);
+
+		uint8_t lo = mmu->directRead(start + index * 16 + lineY * 2);
+		uint8_t hi = mmu->directRead(start + index * 16 + lineY * 2 + 1);
+
+		// Add 8 consecutive horizontal pixels to the line
+		for (int j = 0; j < 8; j++) {
+			uint32_t paletteIndex = bgp[((hi >> (7 - j)) << 1 & 0x02) | ((lo >> (7 - j)) & 0x01)];
+			buffer[i * 8 + j] = palette[paletteIndex];
+		}
+	}
+}
+
 void PPU::updateScanline() {
 	// Early out if scanline is greater than 143, shouldn't ever hit this point
 	if (ly > 143) {
 		return;
 	}
 
-	// If lcdc5 = 1, window is enabled
-	bool lcdc5 = mmu->directRead(0xFF40) & (1 << 5);
+	uint8_t lcdc = mmu->directRead(0xFF40);
 
-	// If lcdc1 = 1, objects are enabled
-	bool lcdc1 = mmu->directRead(0xFF40) & (1 << 1);
+	bool lcdc5 = lcdc & (1 << 5);				// If lcdc5 = 1, window is enabled
+	bool lcdc1 = lcdc & (1 << 1);				// If lcdc1 = 1, objects are enabled
+	bool lcdc0 = lcdc & (1 << 0);				// If lcdc0 = 1, background and window are enabled
+												// Otherwise they are white (palette[0])
+	bool lcdc4 = lcdc & (1 << 4);				// If lcdc4 = 1, 0-127 starts at 0x8000 and 128-255 starts at 0x8800
+												// If lcdc4 = 0, 0-127 starts at 0x9000 and 128-255 starts at 0x8800
+	bool lcdc3 = lcdc & (1 << 3);				// If lcdc3 = 1, the background map starts at 0x9C00, otherwise 0x9800
+	bool lcdc6 = lcdc & (1 << 6);				// If lcdc6 = 1, win map starts at 0x9C00, otherwise 0x9800
 
-	// If lcdc0 = 1, background and window are enabled
-	// Otherwise they are white (palette[0])
-	bool lcdc0 = mmu->directRead(0xFF40) & (1 << 0);
+	firstHalfStart = lcdc4 ? 0x8000 : 0x9000;	
+	secondHalfStart = 0x8800;	
+	backgroundStart = lcdc3 ? 0x9C00 : 0x9800;
+	windowStart = lcdc6 ? 0x9C00 : 0x9800;
 
-	uint8_t scx = mmu->directRead(0xFF43);
+	// Update palette information
+	uint32_t bgpData = mmu->directRead(0xFF47);
+	bgp[0] = (bgpData & 0x03) >> 0;
+	bgp[1] = (bgpData & 0x0C) >> 2;
+	bgp[2] = (bgpData & 0x30) >> 4;
+	bgp[3] = (bgpData & 0xC0) >> 6;
+
+	// TODO: Investigate why performing these unneeded reads seems to improve performance by a couple fps on vita
+	/*uint8_t scx = mmu->directRead(0xFF43);
```

Comment:
Another strange spot I found in perf testing, for some reason uncommenting this makes vita fps from like 35-36 to 37-38. Not sure why, maybe a cache thing or something.